### PR TITLE
[CoSim] specify num threads for Kratos solvers

### PR DESCRIPTION
**Pull request by philbucher**
**Description**
This PR is a **proof of concept** but I think it can be useful.

Usually we only run one type of analysis, and for that one we can figure out the optimal number of omp-threads. 
For CoSimulation however it might be the case that one solver runs faster with a certain number of threads than another one. Hence for those situations it would be necessary to specify different number of threads for each solver, which up to now is not possible

I added a context manager to the Kratos solvers that automatically sets and resets the number of omp-threads before and after the solver is called.
For now I only applied this to the Kratos solvers but this concept can easily be used also for other solvers

**Changelog**
Added the `OpenMPThreadManager` to the Kratos Solvers to control the number of threads the solver runs with

**Additional info**
An alternative approach would be to wrap the functions on a higher level (i.e. the `SolverWrapper`) but I am not sure if this is better since the individual solvers are different after all, so I thought it might be better to do it individually

### CODE REVIEW

File: applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py [hunk 1]
```diff
@@ -10,6 +10,24 @@
...
+class OpenMPThreadManager(object):
+    def __init__(self, num_threads=None):
+        self.num_threads = num_threads
+        if self.num_threads:
+            self.num_threads_orig = KM.OpenMPUtils().GetNumThreads()
+
+    def __enter__(self):
+        if self.num_threads:
+            # print("  >> setting num threads to:", self.num_threads)
+            KM.OpenMPUtils().SetNumThreads(self.num_threads)
```
Comment (roigcarlo) [inline applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py]: I would change this for `KM.OpenMPUtils().SetNumThreads(min(self.num_threads, self.num_threads_orig))`. Otherwise you may be requesting more resources than you have, own, or the user allow you to use.

**Comment by adityaghantasala** [comment]
I quite like the idea. Can you make that small change that @roigcarlo  suggested ? And then I think it will be ready to merge.

File: applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py [hunk 2]
```diff
@@ -30,34 +48,49 @@ def __init__(self, settings, solver_name):
         # this creates the AnalysisStage, creates the MainModelParts and allocates the historial Variables on the MainModelParts:
         self._analysis_stage = self.__GetAnalysisStage()
 
+        if self.settings["solver_wrapper_settings"].Has("omp_num_threads"):
+            omp_num_threads = self.settings["solver_wrapper_settings"]["omp_num_threads"].GetInt()
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py]: there is also a setting for specifying the number of threads in ProjectParameters.json no ?

Comment (adityaghantasala) [reply applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py]: I mean of individual solvers (for kratos in general )?

Comment (philbucher) [reply applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py]: no there is nothing like that yet

From what I understand so far the user had to explicitly set `OMP_NUM_THREADS`

Comment (adityaghantasala) [reply applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py]: ok .. Then perfect !

Comment (adityaghantasala) [reply applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py]: By the way on a side note, having such a setting would we good. Just like here.

**Comment by RiccardoRossi** [comment]
mmm,

settijg the threads within kratos might interfere with task managment
systme,

@rtosi, @mnunez can u comment

**Comment by philbucher** [comment]
@marcnunezc @riccardotosi

**Comment by marcnunezc** [comment]
> mmm, settijg the threads within kratos might interfere with task managment systme, @RTOSI, @mnunez can u comment

You tagged wrong users :sweat_smile: 

What we have been using so far is the python module `threadpool_tcl`. 

https://github.com/KratosMultiphysics/Kratos/blob/1b7bbf71a39717ce2b55a153bdb8e86d005dcece/applications/MultilevelMonteCarloApplication/python_scripts/mlmc_utilities.py#L199-L213

I dont quite remember the reason we started using this, probably @riccardotosi will know better.

I think at first we were trying to set the OMP_THREADS for each task using the @constraint() decorator from pyCOMPS and there were some conflicts there with the OMP_NUM_THREADS env variable (meaning that the actual Kratos simulation was not being run with the constraint that was set).

If we can control this from Kratos itself that would be nice, maybe we can try the approach in this PR with xMC and pyCOMPSS @riccardotosi .

**Comment by roigcarlo** [comment]
I would advice against replacing `threadpool_ctl` for our implementation. Mainly because `OpenMPThreadManager` targets, as its name suggest, OpenMP. No intel komp, mkl, blas, etc... with all those libs having a different way to set the limits of available processors. The reason BSC suggested to manage the threads with that lib is that at some extend it acts as a wrapper setting the limits for the most important libraries out there I would bet that PyCOMPSs works with komp but I am not 100% sure (that's why kratos was ignoring pycompss constraints). 
You can check that from https://github.com/joblib/threadpoolctl/blob/master/threadpoolctl.py#L165

Having that said, `threadpool_ctl` should not interfere with the behavior of `OpenMPThreadManager`, as the later will ask for the current, limit, not the max limit. Moreover, If we were to replace that at XMC or task level (or maybe even at cosimulation level before calling a non-Kratos solver) we could be setting a value that the solver/library below would ignore.

Also @RiccardoRossi the scenario you suggest should be covered by the first comment I made, with that change we are honoring the possible limit made at `threadpool_ctl` level. I would like to test it but it should work.

**TL;DR**:
- `threadpool_ctl` & `OpenMPThreadManager` should be compatible fork tasks
- We should not replace `threadpool_ctl` with `OpenMPThreadManager`
- `OpenMPThreadManager` should respect `threadpool_ctl` limits if specified.

**Comment by philbucher** [comment]
Thx guys for your thoughts!

I will apply your suggestions and give it a try :)

**Comment by philbucher** [comment]
Once #7474 is done, there will be some minor changes

**Review by roigcarlo — APPROVED** [review]
